### DISCUSSION TITLE
Java release type setting not honored?

### DISCUSSION THREAD

**Discussion by bric3** [jdx/mise#2991] [Q&A]
Hi I'm new to `mise`, I followed the setup instructions correctly I believe:

In particular for the `release_type`: https://mise.jdx.dev/lang/java.html#tool-options

So I added in `.config/mise/config.toml` the following settings

```
[tools]
"java" = { version = "corretto-23", release_type = "ea" }
```

Corretto 23 is correctly setup:

```shell
$ mise ls java
Tool  Version                Config Source              Requested
java  corretto-8.432.06.1
java  corretto-11.0.25.9.1
java  corretto-17.0.13.11.1
java  corretto-19.0.2.7.1
java  corretto-21.0.4.7.1
java  corretto-21.0.5.11.1
java  corretto-23.0.1.8.1    ~/.config/mise/config.toml corretto-23
```

However, when listing java distributions I expected to be able to see early access releases, e.g. those of https://jdk.java.net/24/, or more verbosely those that are here : https://github.com/halcyon/asdf-java/blob/049593c66bc287da4a7ac07cab8f916408366c5c/data/jdk-macosx-aarch64-ea.tsv.

```shell
$ mise ls-remote java
...
```

However they are not listed there.

Did I miss something ?

**Comment by roele** [comment]
I'm afraid there are multiple issues with this feature. Generally installing EA versions should work but there is no way at the moment to (remote) list these.

1) Early Access releases are unfortunately not always marked properly in the meta data we use (https://github.com/jdx/rtx-java-metadata). For example Corretto releases are all marked as `ga` even though some releases are defined as "pre-release" in GitHub (e.g. 23.0.0.35.1) so they're listed as GA with the `ls-remote` command.

2) The file for `ea` releases ~is~ was not available. We might have (accidentally?) removed that when we were optimizing the version repository for `mise` a while back. (i re-added the `ea` versions now).
For example: https://rtx-java-metadata.jdx.dev/metadata/ea/macosx/aarch64.json

3) Download URLs are not working anymore. This is especially true for OpenJDK EA releases, once a versions becomes GA the (old) EA URLs wont work anymore (metadata is not updated accordingly).
For example: https://download.java.net/java/early_access/jdk24/9/GPL/openjdk-24-ea+9_macos-aarch64_bin.tar.gz

3) The `ls` and `ls-remote` command are currently not aware of options, so its not possible to list EA versions. Other backends such as `ubi` allow to specify options such as `mise ls-remote ubi:gitlab-org/cli[exe=glab,provider=gitlab]@latest` unfortunately those do not work for core plugins.

**Reply by bric3** [reply]
Oh that is sad.

Also is there a reason for this fork https://github.com/jdx/rtx-java-metadata, instead of the original metadata site by @joschi ?

**Reply by roele** [reply]
Yes. Mainly decoupling, so we can decide what features/fixes go in etc. We also only use a subset of the metadata since `mise` only supports certain OS/Arch types. We also don't use all properties currently provided and wanted to save some bandwidth when downloading these large JSON files.

**Reply by bric3** [reply]
> The `ls` and `ls-remote` command are currently not aware of options, so its not possible to list EA versions. Other backends such as `ubi` allow to specify options such as `mise ls-remote ubi:gitlab-org/cli[exe=glab,provider=gitlab]@latest` unfortunately those do not work for core plugins.

Should an issue be opened for this ?

**Comment by roele** [comment]
Having said that the following will work in terms of (just) installing...

```
[tools]
"java" = { version = "openjdk-25.0.0-ea+18", release_type = "ea" }
```

**Reply by bric3** [reply]
Yes seems unpractical to install specific "EA" variant, like [_loom_](https://jdk.java.net/loom/), [_valhalla_](https://jdk.java.net/valhalla/), or [_leyden_](https://jdk.java.net/leyden/)

**Reply by roele** [reply]
Should work for those too though
```
[tools]
"java" = { version = "openjdk-25.0.0-loom+1-11", release_type = "ea" }
```

**Reply by bric3** [reply]
but then how to use it specifically in a folder or shell ? If this another JDK is running.

**Reply by roele** [reply]
What do you mean exactly? Are you familiar with the `mise.toml` and its inheritance concept?

You basically have "global" configuration in ~/.config/mise/config.toml which for example has `"java" = { version = "temurin-21" }`. When you open a shell and type `java -version` this should show Temurin 21.x as the global config is used.

Then you have a folder in `~/Downloads/my-project` with a `mise.toml` which has `"java" = { version = "openjdk-25.0.0-loom+1-11", release_type = "ea" }`. Once you `cd` into this folder, `mise` puts the EA JDK onto the PATH and `java -version` should show OpenJDK 25 Loom.

**Reply by bric3** [reply]
Thank you! I wasn't aware of that, I was using _asdf_ files like `.tool-versions`.
The workaround you suggested works! E.g. indeed after creating the local file, like you suggested:

```toml
[tools]
"java" = { version = "openjdk-25.0.0-loom+1-11", release_type = "ea" }
```

I had to _trust_ and _install_.

```
❯ mise trust
mise trusted /Users/brice.dutheil/tmp
mise WARN  unknown field in ~/tmp/mise.toml: java
❯ vim mise.toml
mise WARN  missing: java@openjdk-25.0.0-loom+1-11
❯
❯ mise install
mise To enable macOS integration, run the following commands:
sudo mkdir /Library/Java/JavaVirtualMachines/openjdk-25.0.0-loom+1-11.jdk
sudo ln -s /Users/brice.dutheil/.local/share/mise/installs/java/openjdk-25.0.0-loom+1-11/Contents /Library/Java/JavaVirtualMachines/openjdk-25.0.0-loom+1-11.jdk/Contents

openjdk version "25-loom" 2025-09-16
OpenJDK Runtime Environment (build 25-loom+1-11)
OpenJDK 64-Bit Server VM (build 25-loom+1-11, mixed mode, sharing)
mise java@openjdk-25.0.0-loom+1-11 ✓ installed
❯ java --version
openjdk 25-loom 2025-09-16
OpenJDK Runtime Environment (build 25-loom+1-11)
OpenJDK 64-Bit Server VM (build 25-loom+1-11, mixed mode, sharing)
```

That said it would be better to have a proper `ea` support via cli commands, like `ls-remote` for "discovery" and `use` for easier usage.

**Reply by roele** [reply]
Good news is that this also works with the `use` command which takes tool version options in bracket notation.

```bash
mise use java[release_type=ea]@openjdk-25
```

**Reply by bric3** [reply]
Oh nice! (Note one may need to escape the brackets in some shells, e.g. ZSH complains about `no match found` so the command to run is `mise use java\[release_type=ea\]@openjdk-25`)

Now commands like `ls-remote` may need to support that.

Thank you

**Reply by bric3** [reply]
Additional note, install works too

```
mise java\[release_type=ea\]@opendjdk-25.0.0-ea+19`
```

However, for EA distributions after they are installed, `mise use java\[release_type=ea\]@openjdk-25` doesn't pick the regular early access openjdk, it picks the loom variant.

<img width="889" alt="image" src="https://github.com/user-attachments/assets/5f173f7f-937f-4438-9cc0-6974e43877eb" />

One need to specify the version explicitly, and it needs to be the `version` not the `java_version`.

<img width="534" alt="image" src="https://github.com/user-attachments/assets/9114ce39-0d7c-4307-8f11-43bf960bb2b0" />

I wonder if this could make sense to support variant classifier, to avoid maintaining versions e.g.

```
# For regular ea
mise use java\[release_type=ea\]@openjdk-25-ea

# For loom variants
mise use java\[release_type=ea\]@openjdk-25-loom
```

**Reply by bric3** [reply]
FYI I added this syntax in https://github.com/tldr-pages/tldr/pull/16282, but I believe this should be mentioned in the man page.

**Reply by koppor** [reply]
Can someone guide me trying this on Windows, git bash?

```
$ mise use java[release_type=ea]@openjdk-25
mise ERROR failed to install core:java@openjdk-25
mise ERROR no metadata found for version openjdk-25
mise ERROR Run with --verbose or MISE_VERBOSE=1 for more information
```

**Reply by roele** [reply]
There are remaining issues with EA versions for the Java plugin because of caching etc. You might need to clear your caches via `mise cache clear` before trying to install/use a EA version.

The a `openjdk-25` version should be available as per data
```
curl https://rtx-java-metadata.jdx.dev/metadata/ea/windows/x86_64.json | jq '.[] | select(.vendor == "openjdk") | .version'
```
